### ISSUE TITLE
Evolutionha

### ISSUE THREAD

**Issue by dalanmaobu** [zju3dv/snake#163]
`class Evolution(nn.Module):
    def __init__(self):
        super(Evolution, self).__init__()

        self.fuse = nn.Conv1d(128, 64, 1)                                                                                                                                                                                                  
        self.init_gcn = Snake(state_dim=128, feature_dim=64+2, conv_type='dgrid')
        self.evolve_gcn = Snake(state_dim=128, feature_dim=64+2, conv_type='dgrid')
        self.iter = 2
        for i in range(self.iter):
            evolve_gcn = Snake(state_dim=128, feature_dim=64+2, conv_type='dgrid')
            self.__setattr__('evolve_gcn'+str(i), evolve_gcn)`

作者你好，我刚开始看您的代码，有一些还不太明白，能问一下这里的self.fuse()，self.init_gcn，self.evolve_gcn()所实现的具体功能吗？还有feature_dim为什么等于64+2呢？

**Comment by pengsida** [comment]
建议学习一下pytorch。

**Comment by dalanmaobu** [comment]
作者您好，非常感谢您能提供这么优秀的算法，对于论文中提到的我还有几处没有看明白，希望占用您一点宝贵的时间。
1、对于初始轮廓，为什么要从菱形的顶点向物体的极值点用deep snake呢？这个极值点是怎么得到的呢？
2、论文中，有这么一句话：In practice, to consider more context information, the diamond contour is uniformly upsampled to 40 points, and&nbsp;deep snake correspondingly outputs 40 offsets.这里的40是指在菱形块上的采样点吗？那么下面所说的N=128的采样点，和这个有什么区别呢？
3、


代码中，x&nbsp;=&nbsp;self.head(x,&nbsp;adj)，这个head的作用是什么呢？


我的问题可能有些幼稚，但您的回答将对我受益匪浅，再次感谢您的优秀工作，还望能够为我解答一下，谢谢。
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zju3dv/snake"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年3月22日(星期一) 中午12:53
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zju3dv/snake] Evolutionha (#163)





 
Closed #163.

**Comment by dalanmaobu** [comment]
作者您好，非常感谢您能提供这么优秀的算法，对于论文中提到的我还有几处没有看明白，希望占用您一点宝贵的时间。
1、对于初始轮廓，为什么要从菱形的顶点向物体的极值点用deep snake呢？这个极值点是怎么得到的呢？
2、论文中，有这么一句话：In practice, to consider more context information, the diamond contour is uniformly upsampled to 40 points, and deep snake correspondingly outputs 40 offsets.这里的40是指在菱形块上的采样点吗？那么下面所说的N=128的采样点，和这个有什么区别呢？
3、


代码中，x = self.head(x, adj)，这个head的作用是什么呢？这里X变量输出的结果中，每一项都是什么含义呢？


我的问题可能有些幼稚，但您的回答将对我受益匪浅，再次感谢您的优秀工作，还望能够为我解答一下，谢谢。
------------------ 原始邮件 ------------------
发件人:                                                                                                                        "zju3dv/snake"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年3月22日(星期一) 中午12:53
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zju3dv/snake] Evolutionha (#163)





 
Closed #163.

**Comment by pengsida** [comment]
1. 极值点用deep snake预测得到的。
2. 对的，40是指在菱形块上的采样点，N=128是octagon contour上的采样点。
3. head是网络的一部分，head的输出是intermediate feature。

**Comment by dalanmaobu** [comment]
您好，非常您在百忙之中回复我的问题，我还是有几个问题想要请教一下，问题如下：
这处代码我不是很理解，对于这个cnn_feature的数据处理过程我没太看懂，以及传入snake的init_input参数，所代表的的是什么呢？可否麻烦您详解的为我解答一下？



以及i_it_poly、c_it_poly、ind的含义，能否也解释一下呢？
再次感谢，望您可以为我解惑。
def&nbsp;evolve_poly(self,&nbsp;snake,&nbsp;cnn_feature,&nbsp;i_it_poly,&nbsp;c_it_poly,&nbsp;ind):
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;len(i_it_poly)&nbsp;==&nbsp;0:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;torch.zeros_like(i_it_poly)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;h,&nbsp;w&nbsp;=&nbsp;cnn_feature.size(2),&nbsp;cnn_feature.size(3)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;init_feature&nbsp;=&nbsp;snake_gcn_utils.get_gcn_feature(cnn_feature,&nbsp;i_it_poly,&nbsp;ind,&nbsp;h,&nbsp;w)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;c_it_poly&nbsp;=&nbsp;c_it_poly&nbsp;*&nbsp;snake_config.ro
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;init_input&nbsp;=&nbsp;torch.cat([init_feature,&nbsp;c_it_poly.permute(0,&nbsp;2,&nbsp;1)],&nbsp;dim=1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;adj&nbsp;=&nbsp;snake_gcn_utils.get_adj_ind(snake_config.adj_num,&nbsp;init_input.size(2),&nbsp;init_input.device)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;i_poly&nbsp;=&nbsp;i_it_poly&nbsp;*&nbsp;snake_config.ro&nbsp;+&nbsp;snake(init_input,&nbsp;adj).permute(0,&nbsp;2,&nbsp;1)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;i_poly




def&nbsp;get_gcn_feature(cnn_feature,&nbsp;img_poly,&nbsp;ind,&nbsp;h,&nbsp;w):
&nbsp;&nbsp;&nbsp;&nbsp;img_poly&nbsp;=&nbsp;img_poly.clone()
&nbsp;&nbsp;&nbsp;&nbsp;img_poly[...,&nbsp;0]&nbsp;=&nbsp;img_poly[...,&nbsp;0]&nbsp;/&nbsp;(w&nbsp;/&nbsp;2.)&nbsp;-&nbsp;1
&nbsp;&nbsp;&nbsp;&nbsp;img_poly[...,&nbsp;1]&nbsp;=&nbsp;img_poly[...,&nbsp;1]&nbsp;/&nbsp;(h&nbsp;/&nbsp;2.)&nbsp;-&nbsp;1
&nbsp;&nbsp;&nbsp;&nbsp;batch_size&nbsp;=&nbsp;cnn_feature.size(0)
&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;gcn_feature&nbsp;=&nbsp;torch.zeros([img_poly.size(0),&nbsp;cnn_feature.size(1),&nbsp;img_poly.size(1)]).to(img_poly.device)
&nbsp;&nbsp;&nbsp;&nbsp;for&nbsp;i&nbsp;in&nbsp;range(batch_size):
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;poly&nbsp;=&nbsp;img_poly[ind&nbsp;==&nbsp;i].unsqueeze(0)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;feature&nbsp;=&nbsp;torch.nn.functional.grid_sample(cnn_feature[i:i+1],&nbsp;poly)[0].permute(1,&nbsp;0,&nbsp;2)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;gcn_feature[ind&nbsp;==&nbsp;i]&nbsp;=&nbsp;feature

&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;gcn_feature



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zju3dv/snake"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年3月24日(星期三) 中午11:43
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zju3dv/snake] Evolutionha (#163)





  
极值点用deep snake预测得到的。
 
对的，40是指在菱形块上的采样点，N=128是octagon contour上的采样点。
 
head是网络的一部分，head的输出是intermediate feature。

**Comment by dalanmaobu** [comment]
作者您好，我又看了看您的论文，自己测试的kins数据集，想问一下，为什么在cityscapes和kins的效果不一样呢，感觉在cityscapes中处理了遮挡问题，但在kins中，实例之间的遮挡问题，前后覆盖顺序没有处理，有点不解这是什么原因，希望在您有时间的时候可以为我解答一下，谢谢！

**Comment by dalanmaobu** [comment]
作者你好，我想请问一下这个测试数据在代码中是怎么显示出来的？




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "zju3dv/snake"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年3月24日(星期三) 中午11:43
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [zju3dv/snake] Evolutionha (#163)





  
极值点用deep snake预测得到的。
 
对的，40是指在菱形块上的采样点，N=128是octagon contour上的采样点。
 
head是网络的一部分，head的输出是intermediate feature。
